src/Storages/StorageURL.cpp

```diff
@@ -234,6 +236,55 @@ Pipe IStorageURLBase::read(
         chooseCompressionMethod(request_uri.getPath(), compression_method)));
 }
 
+
+Pipe StorageURLWithFailover::read(
+    const Names & column_names,
+    const StorageMetadataPtr & metadata_snapshot,
+    SelectQueryInfo & query_info,
+    ContextPtr local_context,
+    QueryProcessingStage::Enum processed_stage,
+    size_t max_block_size,
+    unsigned /*num_streams*/)
+{
+    auto params = getReadURIParams(column_names, metadata_snapshot, query_info, local_context, processed_stage, max_block_size);
+    for (const auto & uri_option : uri_options)
+    {
+        auto request_uri = uri_option;
+        for (const auto & [param, value] : params)
+            request_uri.addQueryParameter(param, value);
+        try
+        {
+            /// Check for uri accessibility is done in constructor of ReadWriteBufferFromHTTP while creating StorageURLSource.
+            auto url_source =  std::make_shared<StorageURLSource>(
+                request_uri,
+                getReadMethod(),
+                getReadPOSTDataCallback(
+                    column_names, metadata_snapshot, query_info,
+                    local_context, processed_stage, max_block_size),
+                format_name,
+                format_settings,
+                getName(),
+                getHeaderBlock(column_names, metadata_snapshot),
+                local_context,
+                metadata_snapshot->getColumns(),
+                max_block_size,
+                ConnectionTimeouts::getHTTPTimeouts(local_context),
+                chooseCompressionMethod(request_uri.getPath(), compression_method));
+
+            std::shuffle(uri_options.begin(), uri_options.end(), thread_local_rng);
+
+            return Pipe(url_source);
+        }
+        catch (...)
+        {
+            tryLogCurrentException(__PRETTY_FUNCTION__);
```

Comment:
Let's show this to the user so that they don't have to go to the log to debug the problem. Similar to ConnectionPoolWithFailover when it shows "All connection tries failed".